src/Extension/RuleContext.cs

```diff
@@ -1,23 +1,31 @@
-﻿namespace Extension
+﻿using Microsoft.DotNet.Interactive.Events;
+using System.Collections.Generic;
+
+namespace Extension
 {
     public class RuleContext
     {
         public string Name { get; set; }
-        public Challenge Challenge { get; private set; }
+        public Challenge Challenge => _challenge;
         public bool Passed { get; private set; } // do we need this?
+        public string SubmittedCode => _submission.SubmittedCode;
+        public IEnumerable<KernelEvent> EventsProduced => _submission.EventsProduced;
 
         private Evaluation _evaluation;
+        private readonly Challenge _challenge;
+        private readonly ChallengeSubmission _submission;
 
         public RuleContext()
         {
 
         }
 
-        public RuleContext(Challenge challenge, Evaluation evaluation, string defaultName = "")
+        internal RuleContext(Challenge challenge, Evaluation evaluation, string defaultName = "")
```

Comment:
What is the evaluation we pass here? is it the challengeEvaluation? we need to change this. I think that a RuleContext should have access to the  current ChallengeContext and not to an evaluation. That would give access to challenge and all without having to share object reference ownership

Reply:
Ive changed to take a challenge context as a parameter. ChallengeEvaluation will be a computed value.